### PR TITLE
Run3-hcx353 Merging the changes of ZDC code from Anna which was used by the ZDC collaboration for the Run2 setup

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Merging the changes of ZDC code from Anna which was used by the ZDC collaboration for the Run2 setup

#### PR validation:

Use the runTheMatrix test workflows

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

The new XML for ZDC should go for the Run2 scenario definition.

### CODE REVIEW

File: SimG4CMS/Forward/src/ZdcNumberingScheme.cc [hunk 1]
```diff
@@ -5,162 +5,170 @@
...
-
-      for (int ich = 0; ich < level; ich++) {
-        if (name[ich] == "ZDC") {
-          if (copyno[ich] == 1)
-            zside = 1;
-          if (copyno[ich] == 2)
-            zside = -1;
-        } else if (name[ich] == "ZDC_EMLayer") {
-          section = HcalZDCDetId::EM;
+#define EDM_ML_DEBUG
```
Comment (iarspider) [inline SimG4CMS/Forward/src/ZdcNumberingScheme.cc:13]: Please comment this line

File: SimG4CMS/Forward/src/ZdcShowerLibrary.cc [hunk 2]
```diff
@@ -243,3 +270,13 @@ int ZdcShowerLibrary::photonFluctuation(double eav, double esig, double edis) {
     nphot = 0;
   return nphot;
 }
+
+int ZdcShowerLibrary::encodePartID(G4int parCode) {
```
Comment (iarspider) [inline SimG4CMS/Forward/src/ZdcShowerLibrary.cc:274]:
```suggestion
int ZdcShowerLibrary::encodePartID(G4int parCode) {
  return (parCode == emPDG || parCode == epPDG || parCode == gammaPDG) ? 0 : 1;
}
```

File: SimTransport/HectorProducer/src/Hector.cc [hunk 3]
```diff
@@ -86,13 +69,12 @@ Hector::Hector(const edm::ParameterSet &param,
...
                                 << " lengthd1         = " << lengthd1 << "\n"
-                                << "==================================================================="
-                                   "\n";
+                                << "===================================================================\n";
   }
   edm::FileInPath b1(beam1filename.c_str());
   edm::FileInPath b2(beam2filename.c_str());
 
-  // construct beam line for FP420: .
+  // construct beam line for FP420:                                                                                           .
```
Comment (iarspider) [inline SimTransport/HectorProducer/src/Hector.cc:77]: Is this change intentional?

File: SimTransport/HectorProducer/src/Hector.cc [hunk 4]
```diff
@@ -104,11 +86,11 @@ Hector::Hector(const edm::ParameterSet &param,
     m_beamlineFP4202->calcMatrix();
   } else {
     if (m_verbosity)
-      LogDebug("HectorSetup") << "Hector: WARNING: lengthfp420=  " << lengthfp420;
+      edm::LogVerbatim("HectorSetup") << "Hector: WARNING: lengthfp420=  " << lengthfp420;
   }
 
   if (m_ZDCTransport && lengthzdc > 0. && lengthd1 > 0.) {
-    // construct beam line for ZDC: .
+    // construct beam line for ZDC:                                                                                           .
```
Comment (iarspider) [inline SimTransport/HectorProducer/src/Hector.cc:93]: Is this change intentional?

File: SimTransport/HectorProducer/src/Hector.cc [hunk 5]
```diff
@@ -118,7 +100,7 @@ Hector::Hector(const edm::ParameterSet &param,
     m_beamlineZDC1->calcMatrix();
     m_beamlineZDC2->calcMatrix();
 
-    // construct beam line for D1: .
+    // construct beam line for D1:                                                                                           .
```
Comment (iarspider) [inline SimTransport/HectorProducer/src/Hector.cc:103]: Is this change intentional?

File: SimTransport/HectorProducer/src/Hector.cc [hunk 6]
```diff
@@ -164,8 +146,19 @@ void Hector::clear() {
   m_isCharged.clear();
 }
 
-void Hector::add(const HepMC::GenEvent *evt, const edm::EventSetup &iSetup) {
-  H_BeamParticle *h_p;
+/*
```
Comment (iarspider) [inline SimTransport/HectorProducer/src/Hector.cc:149]: Please remove commented out code

File: SimTransport/HectorProducer/src/Hector.cc [hunk 7]
```diff
@@ -366,46 +358,43 @@ void Hector::filterZDC(TRandom3 *rootEngine) {
...
       }
       // if stopfp420 charged particles
       /*
-        else if ( ((*m_isCharged.find(line)).second) ){
-        m_isStoppedzdc[line] = false;// not stopped in propagating to FP420 and
-        therefore in  propagation to ZDC too. if(m_verbosity)
-        LogDebug("HectorEventProcessing") << "Hector:filterZDC: barcode = " <<
-        line << " isStopped=" << (*m_isStoppedzdc.find(line)).second;
-        } */
+	else if ( ((*m_isCharged.find(line)).second) ){
```
Comment (iarspider) [inline SimTransport/HectorProducer/src/Hector.cc:379]: Please remove commented code

File: SimG4CMS/Forward/src/ZdcNumberingScheme.cc [hunk 8]
```diff
@@ -5,162 +5,170 @@
...
+    intindex = packZdcIndex(section, layer, fiber, channel, zside);
 #endif
 
-      bool true_for_positive_eta = true;
-      if (zside == -1)
-        true_for_positive_eta = false;
-
-      HcalZDCDetId zdcId(section, true_for_positive_eta, channel);
-      index = zdcId.rawId();
+    bool true_for_positive_eta = true;
```
[the remark this answers is not in the document]
Comment (iarspider) [reply SimG4CMS/Forward/src/ZdcNumberingScheme.cc:97]: Is this even needed? I think it is only used once here - https://github.com/cms-sw/cmssw/pull/43161/files#diff-ad76782153f0efc9815cb154ef99be3f9bb309762a685ddd1fec0afa043f2b95R101 .

**Review by iarspider — CHANGES_REQUESTED** [review]

**Comment by bsunanda** [comment]
@iarspider Tried to take care of all the comments

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 9]
```diff
@@ -60,245 +62,337 @@ ZdcSD::ZdcSD(const std::string& name,
...
-
-  if (etrack >= zdcHitEnergyCut) {
-    // create hits only if above threshold
+bool ZdcSD::ProcessHits(G4Step* aStep, G4TouchableHistory*) {
+  NaNTrap(aStep);
 
+  if (aStep == nullptr) {
+    return true;
+  } else {
+    if (useShowerLibrary) {
```
[the remark this answers is not in the document]
Comment (iarspider) [reply SimG4CMS/Forward/src/ZdcSD.cc:80]: Maybe remove the `else` branch, since it spans the whole function body? And please leave a note why the commented out section

```
    if (useShowerLibrary) {
      getFromLibrary(aStep);
    }
```

was left.

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 10]
```diff
@@ -60,245 +62,337 @@ ZdcSD::ZdcSD(const std::string& name,
...
+const double EMIN = 1.75715 /*eV*/;                                    // Minimum energy of PMT sensitivity range
+const double ALPHA = /*1/137=*/0.0072973525693;                        // Fine structure constant
+const double HBARC = 6.582119514E-16 /*eV*s*/ * 2.99792458E8 /*m/s*/;  // hbar * c
 
-  const G4ThreeVector& hitPoint = preStepPoint->GetPosition();
-  const G4ThreeVector& hit_mom = preStepPoint->GetMomentumDirection();
-  G4double stepL = aStep->GetStepLength() / cm;
-  G4double beta = preStepPoint->GetBeta();
-  G4double charge = preStepPoint->GetCharge();
+std::ofstream out("debug2.text");
```
[the remark this answers is not in the document]
Comment (iarspider) [reply SimG4CMS/Forward/src/ZdcSD.cc:135]: @bsunanda

**Review by iarspider — COMMENTED** [review]
Mostly LGTM, left a few additional comments.

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: SimG4CMS/Forward/src/ZdcNumberingScheme.cc [hunk 11]
```diff
@@ -5,162 +5,167 @@
...
-          if (copyno[ich] == 1)
-            zside = 1;
-          if (copyno[ich] == 2)
-            zside = -1;
-        } else if (name[ich] == "ZDC_EMLayer") {
-          section = HcalZDCDetId::EM;
+ZdcNumberingScheme::ZdcNumberingScheme(int iv) {
+  verbosity = iv;
+  if (verbosity > 0)
+    edm::LogVerbatim("ForwardSim") << "Creating ZDCNumberingScheme" << std::endl;
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcNumberingScheme.cc:18]: std::endl is not needed.

File: SimG4CMS/Forward/src/ZdcNumberingScheme.cc [hunk 12]
```diff
@@ -5,162 +5,167 @@
...
+  channel = (idx >> 17) & 7;
+  fiber = (idx >> 9) & 255;
+  layer = (idx >> 2) & 127;
+  section = idx & 3;
 
 #ifdef EDM_ML_DEBUG
-    edm::LogVerbatim("ForwardSim") << "ZDC unpacking: idx:" << idx << " -> section " << section << " layer " << layer
-                                   << " fiber " << fiber << " channel " << channel << " zside " << z;
+  edm::LogVerbatim("ForwardSim") << "ZDC unpacking: idx:" << idx << " -> section " << section << " layer " << layer
+                                 << " fiber " << fiber << " channel " << channel << " zside " << z << std::endl;
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcNumberingScheme.cc:145]: endl is not needed.

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 13]
```diff
@@ -60,245 +62,327 @@ ZdcSD::ZdcSD(const std::string& name,
...
+const double ALPHA = /*1/137=*/0.0072973525693;                        // Fine structure constant
+const double HBARC = 6.582119514E-16 /*eV*s*/ * 2.99792458E8 /*m/s*/;  // hbar * c
 
-  const G4ThreeVector& hitPoint = preStepPoint->GetPosition();
-  const G4ThreeVector& hit_mom = preStepPoint->GetMomentumDirection();
-  G4double stepL = aStep->GetStepLength() / cm;
-  G4double beta = preStepPoint->GetBeta();
-  G4double charge = preStepPoint->GetCharge();
+// Calculate the Cherenkov deposit corresponding to a G4Step
+double ZdcSD::calculateCherenkovDeposit(G4Step* aStep) {
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcSD.cc:134]: Here better to have const G4Step*

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 14]
```diff
@@ -60,245 +62,327 @@ ZdcSD::ZdcSD(const std::string& name,
...
+      return (a * lambda + b) / 100.0;
     }
+  }
 
+  return 0;
+}
+
+// Evaluate a function given by set of datapoints
+// Linear interpolation is used to calculate function value between datapoints
+double ZdcSD::evaluateFunction(const std::vector<double> X, const std::vector<double> Y, double x) {
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcSD.cc:332]: Here better to use const std::vector<double>&

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 15]
```diff
@@ -60,245 +62,327 @@ ZdcSD::ZdcSD(const std::string& name,
...
 
-uint32_t ZdcSD::setDetUnitId(const G4Step* aStep) { return ZdcNumberingScheme::getUnitID(aStep); }
+// Energy (eV) to wavelength (nm) conversion
+double ZdcSD::convertEnergyToWavelength(double energy) { return (1240.0 / energy); }
+
+/////////////////////////////////////////////////////////////////////
+
+uint32_t ZdcSD::setDetUnitId(const G4Step* aStep) {
+  uint32_t returnNumber = 0;
+  if (numberingScheme != nullptr)
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcSD.cc:366]: numberingScheme should be defined, this check is not needed.

File: SimG4CMS/Forward/src/ZdcShowerLibrary.cc [hunk 16]
```diff
@@ -27,10 +26,31 @@ ZdcShowerLibrary::ZdcShowerLibrary(const std::string& name, edm::ParameterSet co
 
 ZdcShowerLibrary::~ZdcShowerLibrary() {}
 
-std::vector<ZdcShowerLibrary::Hit>& ZdcShowerLibrary::getHits(const G4Step* aStep, bool& ok) {
-  const G4StepPoint* preStepPoint = aStep->GetPreStepPoint();
-  const G4StepPoint* postStepPoint = aStep->GetPostStepPoint();
-  const G4Track* track = aStep->GetTrack();
+void ZdcShowerLibrary::initRun(G4ParticleTable* theParticleTable) {
+  G4String parName;
+  emPDG = theParticleTable->FindParticle(parName = "e-")->GetPDGEncoding();
```
Comment (civanch) [inline SimG4CMS/Forward/src/ZdcShowerLibrary.cc:31]: These PDG codes can be taken from PDG directly. The call to theParticleTable is not needed.

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by civanch** [comment]
+1

**Comment by rappoccio** [comment]
+1
